### PR TITLE
Wizard recipe: TropicalGemmC-v0.1.0

### PR DESCRIPTION
**Pull request by ArrogantGao**
This pull request contains a new build recipe I built using the BinaryBuilder.jl wizard:

* Package name: TropicalGemmC
* Version: v0.1.0

@staticfloat please review and merge.

### CODE REVIEW

File: T/TropicalGemmC/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    GitSource("https://github.com/ArrogantGao/TropicalGemm_Cuda.git", "3e592f5ccdb1690844b0988f37701271ecda04fa")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+install_license /usr/share/licenses/MIT
+cd TropicalGemm_Cuda/
+make
+exit
```
Comment (imciner2) [inline T/TropicalGemmC/build_tarballs.jl:19]:
```suggestion
```
Not needed.

File: T/TropicalGemmC/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+sources = [
+    GitSource("https://github.com/ArrogantGao/TropicalGemm_Cuda.git", "3e592f5ccdb1690844b0988f37701271ecda04fa")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+install_license /usr/share/licenses/MIT
+cd TropicalGemm_Cuda/
+make
```
Comment (imciner2) [inline T/TropicalGemmC/build_tarballs.jl:18]:
```suggestion
make -j${nproc} 
```

Comment (imciner2) [inline T/TropicalGemmC/build_tarballs.jl:18]: You are missing the installation of the library.

File: T/TropicalGemmC/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    Platform("i686", "linux"; libc = "musl"),
+    Platform("x86_64", "linux"; libc = "musl"),
+    Platform("aarch64", "linux"; libc = "musl"),
+    Platform("armv6l", "linux"; call_abi = "eabihf", libc = "musl"),
+    Platform("armv7l", "linux"; call_abi = "eabihf", libc = "musl")
+]
+
+
+# The products that we will ensure are always built
+products = Product[
```
Comment (imciner2) [inline T/TropicalGemmC/build_tarballs.jl:40]: You need to define the library product for your library.

File: T/TropicalGemmC/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
```
Comment (imciner2) [inline T/TropicalGemmC/build_tarballs.jl:44]: You are using cuda, and need the cuda compiler, so these must be listed as a dependency. You can pull some examples from other scripts in this repo to see how they do it.

**Review by imciner2 — CHANGES_REQUESTED** [review]

**Comment by ArrogantGao** [comment]
Hi, thank you very much for you help, but I am still a little bit confused.
I checked the repo `XGBoost`, which also need to use `nvcc`, and added the following deps:
```
const YGGDRASIL_DIR = "../.."
include(joinpath(YGGDRASIL_DIR, "fancy_toys.jl"))
include(joinpath(YGGDRASIL_DIR, "platforms", "cuda.jl"))

augment_platform_block = CUDA.augment

versions_to_build = [
    v"12.1",
]

# The products that we will ensure are always built
products = [
    LibraryProduct(["TropicalGemmC"], :libtropicalgemm),
]

platforms = [
    Platform("x86_64", "linux"),
]

for cuda_version in versions_to_build, platform in platforms

    cuda_platform = (os(platform) == "linux") && (arch(platform) in ["x86_64"])
    if !isnothing(cuda_version) && !cuda_platform
        continue
    end
    
    if cuda_platform
        augmented_platform = Platform(arch(platform), os(platform);
            cxxstring_abi = cxxstring_abi(platform),
            cuda=isnothing(cuda_version) ? "none" : CUDA.platform(cuda_version)
        )
    else
        augmented_platform = deepcopy(platform)
    end
    should_build_platform(triplet(augmented_platform)) || continue

    dependencies = AbstractDependency[
        Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae"); platforms=filter(!Sys.isbsd, [augmented_platform])),
    ]

    if !isnothing(cuda_version)
        push!(dependencies, BuildDependency(PackageSpec(name="CUDA_full_jll", version=CUDA.full_version(cuda_version))))
        push!(dependencies, RuntimeDependency(PackageSpec(name="CUDA_Runtime_jll")))
    end

    build_tarballs(ARGS, name, version, sources,  script, [augmented_platform], products, dependencies;
                    preferred_gcc_version=v"8",
                    julia_compat="1.6",
                    augment_platform_block)
end
```

But when I try to run this script, the following error occurred:
```
$ julia build_tarballs.jl 
[ Info: Building for x86_64-linux-gnu-cuda+12.1
ERROR: LoadError: KeyError: key v"12.1.1" not found
Stacktrace:
  [1] getindex
    @ ./dict.jl:484 [inlined]
  [2] resolve_versions!(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, julia_version::Nothing, installed_only::Bool)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/Operations.jl:446
  [3] targeted_resolve(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, preserve::Pkg.Types.PreserveLevel, julia_version::Nothing)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/Operations.jl:1357
  [4] tiered_resolve(env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, julia_version::Nothing, try_all_installed::Bool)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/Operations.jl:1329
  [5] _resolve(io::Base.DevNull, env::Pkg.Types.EnvCache, registries::Vector{Pkg.Registry.RegistryInstance}, pkgs::Vector{Pkg.Types.PackageSpec}, preserve::Pkg.Types.PreserveLevel, julia_version::Nothing)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/Operations.jl:1367
  [6] add(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, new_git::Set{Base.UUID}; preserve::Pkg.Types.PreserveLevel, platform::Platform)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/Operations.jl:1384
  [7] add
    @ ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/Operations.jl:1373 [inlined]
  [8] add(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; preserve::Pkg.Types.PreserveLevel, platform::Platform, kwargs::Base.Pairs{Symbol, Base.DevNull, Tuple{Symbol}, NamedTuple{(:io,), Tuple{Base.DevNull}}})
    @ Pkg.API ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/API.jl:275
  [9] Pkg_add(::Pkg.Types.Context, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:platform, :io), Tuple{Platform, Base.DevNull}}})
    @ BinaryBuilderBase ~/.julia/packages/BinaryBuilderBase/2X19x/src/Prefix.jl:610
 [10] (::BinaryBuilderBase.var"#95#101"{Bool, Prefix, Vector{Pkg.Types.PackageSpec}, Platform, Vector{String}, Vector{String}})()
    @ BinaryBuilderBase ~/.julia/packages/BinaryBuilderBase/2X19x/src/Prefix.jl:678
 [11] activate(f::BinaryBuilderBase.var"#95#101"{Bool, Prefix, Vector{Pkg.Types.PackageSpec}, Platform, Vector{String}, Vector{String}}, new_project::String)
    @ Pkg.API ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/Pkg/src/API.jl:1860
 [12] setup_dependencies(prefix::Prefix, dependencies::Vector{Pkg.Types.PackageSpec}, platform::Platform; verbose::Bool)
    @ BinaryBuilderBase ~/.julia/packages/BinaryBuilderBase/2X19x/src/Prefix.jl:669
 [13] setup_dependencies
    @ ~/.julia/packages/BinaryBuilderBase/2X19x/src/Prefix.jl:639 [inlined]
 [14] (::BinaryBuilder.var"#setup_deps#27")(f::typeof(is_target_dependency), prefix::Prefix, dependencies::Vector{AbstractDependency}, platform::Platform, verbose::Bool)
    @ BinaryBuilder ~/.julia/packages/BinaryBuilder/0CUml/src/AutoBuild.jl:790
 [15] autobuild(dir::AbstractString, src_name::AbstractString, src_version::VersionNumber, sources::Vector{<:AbstractSource}, script::AbstractString, platforms::Vector, products::Vector{<:Product}, dependencies::Vector{<:AbstractDependency}; verbose::Bool, debug::Bool, skip_audit::Bool, ignore_audit_errors::Bool, autofix::Bool, code_dir::Union{Nothing, String}, require_license::Bool, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
    @ BinaryBuilder ~/.julia/packages/BinaryBuilder/0CUml/src/AutoBuild.jl:793
 [16] build_tarballs(ARGS::Any, src_name::Any, src_version::Any, sources::Any, script::Any, platforms::Any, products::Any, dependencies::Any; julia_compat::String, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
    @ BinaryBuilder ~/.julia/packages/BinaryBuilder/0CUml/src/AutoBuild.jl:347
 [17] top-level scope
    @ ~/code/Yggdrasil/T/TropicalGemmC/build_tarballs.jl:73
in expression starting at /home/xuanzhaogao/code/Yggdrasil/T/TropicalGemmC/build_tarballs.jl:47
```
Is that because some other deps needed?

**Comment by tylerjthomas9** [comment]
> $ julia build_tarballs.jl 
> [ Info: Building for x86_64-linux-gnu-cuda+12.1
> ERROR: LoadError: KeyError: key v"12.1.1" not found

What version of Julia are you using to build? I get this same error on Julia `v1.9`. I am not sure why this happens, but I have just been using `v1.7` for `BinaryBuilder.jl`.

Also, from looking at your build, you also need to install your library in the lib folder. Here is a modified build script:

```julia
script = raw"""
cd $WORKSPACE/srcdir
cd TropicalGemm_Cuda/

export CUDA_HOME=${WORKSPACE}/destdir/cuda;
export PATH=$PATH:$CUDA_HOME/bin

make -j${nproc}

install -Dvm 0755 TropicalGemmC.${dlext} ${libdir}/TropicalGemmC.${dlext}
install_license /usr/share/licenses/MIT
"""
```

**Comment by ArrogantGao** [comment]
> > $ julia build_tarballs.jl
> > [ Info: Building for x86_64-linux-gnu-cuda+12.1
> > ERROR: LoadError: KeyError: key v"12.1.1" not found
> 
> What version of Julia are you using to build? I get this same error on Julia `v1.9`. I am not sure why this happens, but I have just been using `v1.7` for `BinaryBuilder.jl`.
> 
> Also, from looking at your build, you also need to install your library in the lib folder. Here is a modified build script:
> 
> ```julia
> script = raw"""
> cd $WORKSPACE/srcdir
> cd TropicalGemm_Cuda/
> 
> export CUDA_HOME=${WORKSPACE}/destdir/cuda;
> export PATH=$PATH:$CUDA_HOME/bin
> 
> make -j${nproc}
> 
> install -Dvm 0755 TropicalGemmC.${dlext} ${libdir}/TropicalGemmC.${dlext}
> install_license /usr/share/licenses/MIT
> """
> ```

Hi, thank you very much for your help. 
I am currently using the Julia `v1.9.3`, here are the detailed version info.
```
Julia Version 1.9.3
Commit bed2cd540a1 (2023-08-24 14:43 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 128 × Intel(R) Xeon(R) Platinum 8378A CPU @ 3.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, icelake-server)
  Threads: 1 on 128 virtual cores
Environment:
  LD_LIBRARY_PATH = /opt/nvidia/hpc_sdk/Linux_x86_64/23.5/compilers/lib:
  JULIA_PKG_SERVER = https://mirrors.ustc.edu.cn/julia
  JULIA_EDITOR = code
  JULIA_NUM_THREADS =
```

```
CUDA runtime 12.2, artifact installation
CUDA driver 12.1
NVIDIA driver 530.30.2

CUDA libraries: 
- CUBLAS: 12.2.5
- CURAND: 10.3.3
- CUFFT: 11.0.8
- CUSOLVER: 11.5.2
- CUSPARSE: 12.1.2
- CUPTI: 20.0.0
- NVML: 12.0.0+530.30.2

Julia packages: 
- CUDA: 5.0.0
- CUDA_Driver_jll: 0.6.0+3
- CUDA_Runtime_jll: 0.9.2+0

Toolchain:
- Julia: 1.9.3
- LLVM: 14.0.6
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5
- Device capability support: sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80, sm_86
```

**Comment by ArrogantGao** [comment]
Hi, @imciner2.. I revised the script and now all tests passed, could you please help me to merge that?
Thanks a lot.

**Pull request closed without merging**